janus-idp/backstage-plugins
# feat(rbac): implement conditional policies feature.

### Referenced issue:

Fixes: janus-idp/backstage-showcase#429

# How to test it:

1. Login like permission admin using github auth.
2. Retrieve admin user token. Copy it to terminal:

export token='...token...value'

3. Let's provide read catalog entity permission
curl -X POST "http://localhost:7007/api/permission/policies" -d '{"entityReference": "group:default/team-a", "permission": "catalog-entity", "policy": "read", "effect":"allow"}' -H "Content-Type: application/json" -H "Authorization: Bearer $token" -v

After that admin user should have ability to read catalogs, but he can see too much. 

4. Let's set a condition to restrict users to viewing only their own components or components owned by members of their group:

# Condition without criteria
curl -X POST "http://localhost:7007/api/permission/conditions" -d '{"result":"CONDITIONAL","pluginId":"catalog","resourceType":"catalog-entity","conditions":{"rule":"IS_ENTITY_OWNER","resourceType":"catalog-entity","params":{"claims":["user:default/andrienkoaleksandr","group:default/team-a"]}}}' -H "Content-Type: application/json" -H "Authorization: Bearer $token" -v

5. We would like to provide ability to see resources with kind "Group" for user. That's will be condition with criteria:

# Condition with criteria: 
curl -X PUT "http://localhost:7007/api/permission/conditions/1" -d '{"result":"CONDITIONAL","pluginId":"catalog","resourceType":"catalog-entity","conditions":{"anyOf":[{"rule":"IS_ENTITY_OWNER","resourceType":"catalog-entity","params":{"claims":["user:default/andrienkoaleksandr","group:default/team-a"]}},{"rule":"IS_ENTITY_KIND","resourceType":"catalog-entity","params":{"kinds":["Group"]}}]}}' -H "Content-Type: application/json" -H "Authorization: Bearer $token" -v

# Also we can check  condition policy by id via REST api:
curl -X GET "http://localhost:7007/api/permission/conditions/1" -H "Content-Type: application/json" -H "Authorization: Bearer $token" -v
